file: Automatically convert large pasted text into a .txt file upload.

userAdityaa opened the pull request:
When a user pastes a large amount of text into the compose box, it will now be automatically converted into a .txt file upload named 'pasted.txt'. This ensures that messages do not exceed the length limit.

Specifically, if the pasted text exceeds 1000 characters and would cause the message to surpass the allowed limit (e.g., a 10K character limit), it is converted into a file. For instance, if there are already 6000 characters in the compose box, a new 5000-character paste will be uploaded as a file instead of being added as text.

Fixes #33107.

**Screenshots of the Changes: Pasted character (Uploading) > Uploaded > Changes see in the UI**


<details>
<summary>Pre-existing text in the compose box, then the user paste text that is converted to a text file: </summary>
 <img width="1123" alt="Screenshot 2025-05-01 at 10 34 19 AM" src="https://github.com/user-attachments/assets/c34778d1-e93d-4795-b6e7-1bce1f6bfbdc" />
<img width="1123" alt="Screenshot 2025-05-01 at 10 35 06 AM" src="https://github.com/user-attachments/assets/fb458092-9f33-47aa-a5a7-e7d1cbbbcbc2" />
<img width="1123" alt="Screenshot 2025-05-01 at 10 34 37 AM" src="https://github.com/user-attachments/assets/dd33b6d1-29ac-4c16-a3d0-f4013fd015bf" />
</details>

<details>
<summary>When user directly paste text that is larger than 10,000 character: </summary>
<img width="1123" alt="Screenshot 2025-05-01 at 10 37 53 AM" src="https://github.com/user-attachments/assets/1f853194-6584-443a-816c-d9bd677b83cf" />
<img width="1123" alt="Screenshot 2025-05-01 at 10 38 04 AM" src="https://github.com/user-attachments/assets/f95aa2fc-c039-42f0-861d-3ea2d1b64b4a" />
<img width="1123" alt="Screenshot 2025-05-01 at 10 38 11 AM" src="https://github.com/user-attachments/assets/20cb0dfe-cb9a-4226-95c2-6037628da927" />
</details>



<details>
<summary>When a user is almost around 10,000 characters but ends up pasting only 500 or an amount below the MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT threshold of 2,000 characters.</summary>
<img width="1123" alt="Screenshot 2025-05-01 at 10 41 46 AM" src="https://github.com/user-attachments/assets/85a322f8-0337-4e29-b4ca-e83b62b8055a" />
<img width="1123" alt="Screenshot 2025-05-01 at 10 42 04 AM" src="https://github.com/user-attachments/assets/557a5ff1-eb1e-40ef-a69a-4b83ba0ce5c5" />
</details>

<details>
<summary>Upload Banner, Upload text in compose, File link in the message Feed and File preview, Edit text area upload and removal</summary>
<img width="832" alt="Screenshot 2025-05-01 at 2 08 34 PM" src="https://github.com/user-attachments/assets/de91fb4e-018d-40aa-befa-915e09fb554e" />
<img width="829" alt="Screenshot 2025-05-01 at 2 09 04 PM" src="https://github.com/user-attachments/assets/9183fe09-baca-4304-8ccb-a7df8a284317" />
<img width="832" alt="Screenshot 2025-05-01 at 2 09 33 PM" src="https://github.com/user-attachments/assets/9bd4287d-7b21-41fe-a968-b3afd7a07bc2" />
<img width="1440" alt="Screenshot 2025-05-01 at 2 11 53 PM" src="https://github.com/user-attachments/assets/f9f8781c-c4f5-49dd-8083-e0858f414ebb" />
<img width="641" alt="Screenshot 2025-05-07 at 8 23 09 PM" src="https://github.com/user-attachments/assets/6bdded28-dbb1-4ab9-81b7-2f0374613427" />
<img width="488" alt="Screenshot 2025-05-07 at 8 23 44 PM" src="https://github.com/user-attachments/assets/b193052c-077d-4236-9912-5113df4ea668" />
</details>



**Pre-existing character in the composition combined with a pasted character.**

https://github.com/user-attachments/assets/bd414608-0d97-4c0b-99a6-f6c60a2436d3



**Directly inserting a lengthy character sequence.**


https://github.com/user-attachments/assets/0ce2b993-9685-4238-aab4-3fd32532e093


**Trying to paste message less than 2000 (MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT)..**

https://github.com/user-attachments/assets/9b7ac215-9efc-42bd-8582-79147d2f226b







<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/upload.ts:
```diff
@@ -270,6 +271,12 @@ export function rewire_upload_files(value: typeof upload_files): void {
+    upload_files(compose_upload_object, compose_config, [pasted_file]);
+}
+
+set_upload_pasted_file_callback(upload_pasted_file);
```
This should be done instead via passing `upload.upload_pasted_file` from `ui_init` into a `copy_and_paste.initialize(...)` method.

There's several other examples of this pattern, and it's important because we never want state to be mutated when importing a module (it makes unit testing hard).

userAdityaa replied:
Updated the code with this approach!

timabbott commented on web/src/copy_and_paste.ts:
```diff
@@ -881,10 +893,32 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
 
     if (clipboardData.getData) {
         const $textarea = $(this);
+        const existing_text_length = $textarea.val()?.length;
```
Use `!` here, rather than where you use it; the types will be better that way.

userAdityaa replied:
Done.

alya commented on web/src/copy_and_paste.ts:
```diff
@@ -881,10 +893,32 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
+        const is_pasted_text_too_large = paste_text_length > realm.max_message_length;
+
+        const is_combined_text_too_large =
+            paste_text_length + existing_text_length! > realm.max_message_length;
```
The PR description says:

> Specifically, if the pasted text exceeds 1000 characters and would cause the message to surpass the allowed limit

userAdityaa replied:
I somehow missed it, now it's included. I have updated the PR description to include the video for this.

userAdityaa commented:
I have included the min_paste for file treatment in the **constant** file, if necessary I will move it to default_settings.

timabbott commented on web/src/compose_paste.ts:
```diff
@@ -545,10 +555,36 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
+            // Convert the pasted text into a .txt file and upload it
+            const pasted_file = create_text_file(trimmed_paste_text, "pasted.txt");
+            upload_pasted_file(pasted_file);
+            // Prevent the default paste behavior
```
These last two comments are largely useless -- you're just repeating what the code very obviously does if you just read the function names and parameters. 

Comments should help with understanding -- so concise explanations of why this is the correct thing to do are great, or things that add insight, like "This replaces the default paste behavior of adding the content to the compose box."

userAdityaa replied:
I've removed the redundant comments and kept only meaningful explanations that provide insight into the implementation. Let me know if any further clarification is needed!

userAdityaa replied:
Certainly! Here are the steps for the change:  

1. **Identify the Import Cycle**  
   - `copy_and_paste.js` previously imported `upload.ts` to use `upload.upload_pasted_file`.  
   - This caused an import cycle, making the module dependent on a state mutation at import time.  

2. **Refactor to Use Dependency Injection**  
   - Instead of importing `upload.ts` inside `copy_and_paste.js`, modify `copy_and_paste` to accept `upload.upload_pasted_file` as a parameter during initialization.  

3. **Modify `copy_and_paste.js`**  
   - Add an `initialize` function that accepts `upload_pasted_file` as an argument and stores it for later use.  
   - Removed any direct references to `upload.upload_pasted_file` with this injected function.  

4. **Update `ui_init.js`**  
   - Pass `upload.upload_pasted_file` to `copy_and_paste.initialize(...)` during the UI setup process.   

5. **Ensure Consistency with Existing Patterns**  
   - This approach follows similar patterns used elsewhere in the codebase, keeping dependencies modular and easier to test.  

6. **Test the Changes**  
   - Verify that `copy_and_paste` can still call `upload.upload_pasted_file` correctly.  
   - Run unit tests to confirm that the import cycle is resolved and the behavior remains unchanged.  

Let me know if you need further refinements!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -14895,6 +14895,16 @@ paths:
+                          in Unicode code points. Clients should use this property rather than
+                          hardcoding field sizes.
+
+                          **Changes**: New in Zulip 10.0 (feature level 332).
```
332 seems wrong as a number? Where's the change in `version.py`? 

Try using the `api_docs/unmerged.d` protocol.

userAdityaa replied:
I will definitely look into it.

userAdityaa replied:
I’ll definitely review previous API changes, but for now, in this PR, we’ve followed the feedback to define the constant globally within the same file rather than setting it up as a server-side constant.

alexmv commented on web/src/compose_paste.ts:
```diff
@@ -545,10 +555,36 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
+            is_paste_large_enough_for_file
+        ) {
+            // Convert the pasted text into a .txt file and upload it
+            const pasted_file = create_text_file(trimmed_paste_text, "pasted.txt");
```
I would argue that if they pasted a huge block, they didn't have sloppy copying -- they want exactly what they pasted, so we shouldn't be using the trimmed version.

userAdityaa replied:
see your point, and I’ve updated the implementation to use the original pasted length instead of a trimmed version. This ensures that users get exactly what they pasted without unintended modifications.

alexmv replied:
This still looks to use the trimmed version?

userAdityaa replied:
Now it's corrected.

userAdityaa replied:
Apologies for the oversight on my part. I have now carefully reviewed all the code, and we’ve ensured that the untrimmed version is being used for conversion to a file instead of the trimmed one. If you have any further feedback or suggestions, please let me know—I’d be happy to make any necessary adjustments!

timabbott commented on web/src/ui_init.js:
```diff
@@ -562,7 +562,10 @@ export function initialize_everything(state_data) {
-    compose_paste.initialize();
+    compose_paste.initialize({
+        min_paste_size_for_file_treatment: upload.MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT,
+        upload_pasted_file: upload.upload_pasted_file,
```
Maybe it should just be `compose_paste.MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT`, the way you have it is quite convoluted.

userAdityaa replied:
1. **Replaced Complex Reference** → Used `compose_paste.MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT` for clarity.  
2. **Scoped with `const`** → Kept it local since it's only used in the same file.  
3. **Future Flexibility** → Can switch to `export const` if needed later.

userAdityaa replied:
### **Summary of Changes in Text Pasting Logic**  

1. **Existing Text Length Calculation**  
   - The current length of the text in the compose box is determined before the paste operation.  

2. **Pasted Text Length Calculation (Untrimmed)**  
   - The length of the pasted text is measured in its original form, without any trimming.  

3. **Key Calculations:**  
   - **`is_pasted_text_too_large`**: Determines if the pasted text alone exceeds the threshold for automatic conversion into a file.  
   - **`is_combined_text_too_large`**: Checks if the combined length of the existing text and the newly pasted content surpasses the maximum allowed text length (`max_text_length`).  

4. **Handling Large Pasted Text:**  
   - **Case 1:** If the pasted text is **2000 characters or more**, it gets converted into a file.  
   - **Case 2:** If the pasted text is **less than 2000 characters**, it remains as plain text and is not converted into a file.  

5. **Global Configuration:**  
   - The **2000-character threshold** is managed using the `MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT` constant, which is globally defined in the `compose_paste.ts` file.

userAdityaa commented:
### Summary of the whole Implementation: 

1. **Check existing text length** → `existing_text_length` stores the current text length in the input field.  

2. **Get pasted text length** → `pasted_text_length` holds the length of the pasted text without trimming.  

3. **Verify if pasted text is too large** → `is_pasted_text_too_large` checks if it exceeds `realm.max_message_length`.  

4. **Check combined text length** → `is_combined_text_too_large` ensures `pasted_text_length + existing_text_length` does not surpass `realm.max_message_length`.  

5. **Determine if pasted text qualifies for file treatment** → `is_paste_large_enough_for_file` checks if it meets `compose_paste.MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT`.  

6. **Convert to file if necessary** → If the pasted or combined text is too large and meets the file size condition, `create_text_file(paste_text, "pasted.txt")` generates a file, and `upload_pasted_file(pasted_file)` uploads it. Default paste behavior is prevented using `event.preventDefault()` and `event.stopPropagation()`.

alya commented:
@kuv2707 Are you up for doing the next round of review on this one?

kuv2707 commented on web/src/compose_paste.ts:
```diff
@@ -550,10 +562,35 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
+            is_paste_large_enough_for_file
+        ) {
+            const pasted_file = create_text_file(paste_text, "pasted.txt");
+            upload_pasted_file(pasted_file);
```
```suggestion
            upload.upload_pasted_file(pasted_file);
```

kuv2707 commented on web/src/ui_init.js:
```diff
@@ -563,7 +563,9 @@ export function initialize_everything(state_data) {
     click_handlers.initialize();
     scheduled_messages_overlay_ui.initialize();
-    compose_paste.initialize();
+    compose_paste.initialize({
```
This can therefore be removed.

userAdityaa replied:
The suggestion was first implemented, but it causes circular dependency, so I have Initialized it in `ui_init.js` for consistency.

kuv2707 replied:
You can refactor some things to avoid circular dependencies. It is not a good idea to make the implementation convoluted just to work around the circular dep. constraint.

(The point of prohibiting circ. deps is to simplify the structuring of modules).

userAdityaa replied:
Check: https://github.com/zulip/zulip/pull/33542#discussion_r1968311717 ,
Still Let me look for some workarounds for this. Thank you for the feedback.

kuv2707 replied:
To be clear, this was the change I suggested, which doesn't seem to cause any circular dependency:

```diff
diff --git a/web/src/compose_paste.ts b/web/src/compose_paste.ts
index 38ea02c5d7..ee1c5cb072 100644
--- a/web/src/compose_paste.ts
+++ b/web/src/compose_paste.ts
@@ -7,10 +7,14 @@ import TurndownService from "turndown";
 
 import * as compose_ui from "./compose_ui.ts";
 import * as hash_util from "./hash_util.ts";
+import {realm} from "./state_data.ts";
 import * as stream_data from "./stream_data.ts";
 import * as topic_link_util from "./topic_link_util.ts";
+import * as upload from "./upload.ts";
 import * as util from "./util.ts";
 
+const MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT = 2000;
+
 declare global {
     // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
     interface HTMLElementTagNameMap {
@@ -606,6 +610,11 @@ export function try_stream_topic_syntax_text(text: string): string | null {
     return syntax_text;
 }
 
+function create_text_file(text: string, filename: string): File {
+    const blob = new Blob([text], {type: "text/plain"});
+    return new File([blob], filename, {type: "text/plain"});
+}
+
 export function paste_handler(this: HTMLTextAreaElement, event: JQuery.TriggeredEvent): void {
     assert(event.originalEvent instanceof ClipboardEvent);
     const clipboardData = event.originalEvent.clipboardData;
@@ -620,10 +629,35 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
 
     if (clipboardData.getData) {
         const $textarea = $(this);
+        const existing_text_length = $textarea.val()!.length;
         const paste_text = clipboardData.getData("text");
         let paste_html = clipboardData.getData("text/html");
         // Trim the paste_text to accommodate sloppy copying
         const trimmed_paste_text = paste_text.trim();
+        const pasted_text_length = paste_text.length;
+
+        const is_pasted_text_too_large = pasted_text_length > realm.max_message_length;
+
+        const is_combined_text_too_large =
+            pasted_text_length + existing_text_length > realm.max_message_length;
+
+        const is_paste_large_enough_for_file =
+            pasted_text_length >= MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT;
+
+        // Convert pasted text to a file if it exceeds size limits
+        // If the pasted or combined text is too large, create a "pasted.txt" file
+        // and upload it instead of inserting the text directly.
+        if (
+            (is_pasted_text_too_large || is_combined_text_too_large) &&
+            is_paste_large_enough_for_file
+        ) {
+            const pasted_file = create_text_file(paste_text, "pasted.txt");
+            upload.upload_pasted_file(pasted_file);
+
+            event.preventDefault();
+            event.stopPropagation();
+            return;
+        }
 
         // Only intervene to generate formatted links when dealing
         // with a URL and a URL-safe range selection.
diff --git a/web/src/upload.ts b/web/src/upload.ts
index 864426888f..c317cc86d2 100644
--- a/web/src/upload.ts
+++ b/web/src/upload.ts
@@ -274,6 +274,10 @@ export function rewire_upload_files(value: typeof upload_files): void {
     upload_files = value;
 }
 
+export function upload_pasted_file(pasted_file: File): void {
+    upload_files(compose_upload_object, compose_config, [pasted_file]);
+}
+
 // Borrowed from tus-js-client code at
 // https://github.com/tus/tus-js-client/blob/ca63ba254ea8766438b9d422f6f94284911f1fa5/lib/index.d.ts#L79
 // The library does not export this type, hence requiring a copy here.

```

kuv2707 replied:
For some reason, `tus` uploads aren't working on my local machine, so can you test if this change has the same behaviour?

userAdityaa replied:
> To be clear, this was the change I suggested, which doesn't seem to cause any circular dependency:
> 
> ```diff
> diff --git a/web/src/compose_paste.ts b/web/src/compose_paste.ts
> index 38ea02c5d7..ee1c5cb072 100644
> --- a/web/src/compose_paste.ts
> +++ b/web/src/compose_paste.ts
> @@ -7,10 +7,14 @@ import TurndownService from "turndown";
>  
>  import * as compose_ui from "./compose_ui.ts";
>  import * as hash_util from "./hash_util.ts";
> +import {realm} from "./state_data.ts";
>  import * as stream_data from "./stream_data.ts";
>  import * as topic_link_util from "./topic_link_util.ts";
> +import * as upload from "./upload.ts";
>  import * as util from "./util.ts";
>  
> +const MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT = 2000;
> +
>  declare global {
>      // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
>      interface HTMLElementTagNameMap {
> @@ -606,6 +610,11 @@ export function try_stream_topic_syntax_text(text: string): string | null {
>      return syntax_text;
>  }
>  
> +function create_text_file(text: string, filename: string): File {
> +    const blob = new Blob([text], {type: "text/plain"});
> +    return new File([blob], filename, {type: "text/plain"});
> +}
> +
>  export function paste_handler(this: HTMLTextAreaElement, event: JQuery.TriggeredEvent): void {
>      assert(event.originalEvent instanceof ClipboardEvent);
>      const clipboardData = event.originalEvent.clipboardData;
> @@ -620,10 +629,35 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
>  
>      if (clipboardData.getData) {
>          const $textarea = $(this);
> +        const existing_text_length = $textarea.val()!.length;
>          const paste_text = clipboardData.getData("text");
>          let paste_html = clipboardData.getData("text/html");
>          // Trim the paste_text to accommodate sloppy copying
>          const trimmed_paste_text = paste_text.trim();
> +        const pasted_text_length = paste_text.length;
> +
> +        const is_pasted_text_too_large = pasted_text_length > realm.max_message_length;
> +
> +        const is_combined_text_too_large =
> +            pasted_text_length + existing_text_length > realm.max_message_length;
> +
> +        const is_paste_large_enough_for_file =
> +            pasted_text_length >= MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT;
> +
> +        // Convert pasted text to a file if it exceeds size limits
> +        // If the pasted or combined text is too large, create a "pasted.txt" file
> +        // and upload it instead of inserting the text directly.
> +        if (
> +            (is_pasted_text_too_large || is_combined_text_too_large) &&
> +            is_paste_large_enough_for_file
> +        ) {
> +            const pasted_file = create_text_file(paste_text, "pasted.txt");
> +            upload.upload_pasted_file(pasted_file);
> +
> +            event.preventDefault();
> +            event.stopPropagation();
> +            return;
> +        }
>  
>          // Only intervene to generate formatted links when dealing
>          // with a URL and a URL-safe range selection.
> diff --git a/web/src/upload.ts b/web/src/upload.ts
> index 864426888f..c317cc86d2 100644
> --- a/web/src/upload.ts
> +++ b/web/src/upload.ts
> @@ -274,6 +274,10 @@ export function rewire_upload_files(value: typeof upload_files): void {
>      upload_files = value;
>  }
>  
> +export function upload_pasted_file(pasted_file: File): void {
> +    upload_files(compose_upload_object, compose_config, [pasted_file]);
> +}
> +
>  // Borrowed from tus-js-client code at
>  // https://github.com/tus/tus-js-client/blob/ca63ba254ea8766438b9d422f6f94284911f1fa5/lib/index.d.ts#L79
>  // The library does not export this type, hence requiring a copy here.
> ```

By importing `upload.ts` in `compose_paste.ts` this is what lint error says: 
```
eslint    |
eslint    | /srv/zulip/web/src/compose_paste.ts
eslint    |   13:1  error  Dependency cycle via ./compose_reply.ts:13  import/no-cycle
eslint    |
eslint    | /srv/zulip/web/src/upload.ts
eslint    |   13:1  error  Dependency cycle via ./compose_paste.ts:10  import/no-cycle
eslint    |
eslint    | ✖ 2 problems (2 errors, 0 warnings)
```

kuv2707 replied:
Oh, there was a config issue in my linter.
Looks like we can't do much about it.

kuv2707 reviewed (commented):
Left a suggestion which would greatly simplify the implementation.

kuv2707 commented on web/src/compose_paste.ts:
```diff
@@ -550,10 +562,35 @@ export function paste_handler(this: HTMLTextAreaElement, event: JQuery.Triggered
+        const is_paste_large_enough_for_file =
+            pasted_text_length >= MINIMUM_PASTE_SIZE_FOR_FILE_TREATMENT;
+
+        // Convert pasted text to a file if it exceeds size limits
```
This line can be removed as it is clear from the code.

userAdityaa replied:
I have removed that line now. Thanks for the suggestion.

kuv2707 reviewed (approved):
Left a minor suggestion; Looks fine to me otherwise.

alya commented:
Let's name the file "pasted text.txt". Is the name marked for translation?

alya commented:
Playing with the PR, I'm not sure if we need a banner to explain what happened. Might be clear enough as-is.

userAdityaa commented:
> Let's name the file "pasted text.txt". Is the name marked for translation?

I have changed it to be "Pasted text.txt" instead and marked that for translation. I have marked **Pasted text** to translate leaving `.txt` as it is. If that is not ok, I will change it accordingly.

userAdityaa commented:
> Playing with the PR, I'm not sure if we need a banner to explain what happened. Might be clear enough as-is.

I think it looks good though, but I can remove that if we want to.

alya commented:
As usual, please add screenshots of the UI changes to the PR description, not just videos.

userAdityaa commented:
> As usual, please add screenshots of the UI changes to the PR description, not just videos.

I've now added all the necessary images that covers  the UI Changes, and I've also updated the video with the newly suggested file name: `Pasted text.txt`

alya commented:
@timabbott Would you like to take a look at this next?

The experience seems OK in manual testing (I didn't try to test the logic), though it felt like pasting was a bit slow. I dunno if this is a regression; it might be just the general state of things since we added fancy pasting.

timabbott commented on web/src/upload.ts:
```diff
@@ -274,6 +274,10 @@ export function rewire_upload_files(value: typeof upload_files): void {
 
+export function upload_pasted_file(pasted_file: File): void {
+    upload_files(compose_upload_object, compose_config, [pasted_file]);
+}
```
The message edit code path seems to not work properly -- I think you need to pass the `$textarea` into this function rather than unconditionally using `compose_config`.

userAdityaa replied:
Yes, I was able to reproduce the issue and have resolved it. The function now checks whether the paste occurred in a `compose` or `edit textarea`, and uploads files based on the appropriate context.

Additionally, there were some UI changes related to pasting in the edit box. I've updated the PR description with relevant screenshots to reflect those changes.

userAdityaa replied:
This is how the UI Looks now for edit paste upload:  
<img width="641" alt="Screenshot 2025-05-07 at 8 23 09 PM" src="https://github.com/user-attachments/assets/f4facc8b-5993-4574-8ac5-055f68ff742f" />

timabbott commented:
I found one bug, otherwise this looks good to merge. Please DM me when you've fixed it.

timabbott commented on web/src/upload.ts:
```diff
@@ -274,6 +274,20 @@ export function rewire_upload_files(value: typeof upload_files): void {
 }
 
+export function upload_pasted_file(pasted_file: File): void {
+    if ($("textarea").hasClass("message_edit_content")) {
```
This code is nonsense -- you need to pass in the actual textarea that the person is pasting into, not do something just select all textareas.

@kuv2707 would you be up for picking up this PR and submitting a cleaned up version of it?
